### PR TITLE
Fix DynamicAreaDefinition not preserving user's requested resolution

### LINKED ISSUE #517: EWA resampling in 1.27 slows down four times than 1.26.1
#### Code Sample, a minimal, complete, and verifiable piece of code
Some regular satpy codes here
```python
area_id = "abc"
projection = some proj4 strings
area_def = create_area_def(area_id=area_id, projection=projection, resolution=250)
scn = scn.resample(area_def, resampler="ewa", weight_delta_max=10, weight_distance_max=1)

```
#### Problem description

To produce a 250m rayleigh-corrected true color scene from MODIS L1B:
1.26.1: 83.054s (with CPU almost fully loaded)
1.27: 321.692s (CPU usage around 20% all the time)

#### Versions of Python, package at hand and relevant dependencies
python: 3.11.3
satpy: 0.42.3-dev
dask: 2023.5.0

### PR DESCRIPTION
**Pull request by djhoese**
Closes the new issue described in #517 where the user points out that they requested a resolution of 250.0 but the resulting `AreaDefinition` does not match that and actually changes between cases. This PR fixes this so in the cases where resolution is provided (and not shape) the resulting area should have `pixel_size_x/pixel_size_y` very very close to this value except for some floating point precision.

 - [x] Closes #517 
 - [x] Tests added 
 - [x] Tests passed 
 - [ ] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

File: pyresample/geometry.py [hunk 1]
```diff
@@ -1131,6 +1131,9 @@ def compute_domain(
             x_resolution, y_resolution = resolution
             width = int(np.rint((corners[2] - corners[0]) * 1.0 / x_resolution + 1))
             height = int(np.rint((corners[3] - corners[1]) * 1.0 / y_resolution + 1))
+            # align corners with pixel resolution
+            corners[2] = corners[0] + (width - 1) * x_resolution
+            corners[3] = corners[1] + (height - 1) * y_resolution
```
Comment (mraspaud) [inline pyresample/geometry.py:1136]: doesn't this rather belong to `_update_corners_for_full_extent`?

**Review by mraspaud — COMMENTED** [review]
LGTM, just one placement question

File: pyresample/geometry.py (hunk 1 above)
Comment (djhoese) [reply pyresample/geometry.py:1136]: Not strictly, but I agree it seems like an odd place to put this code compared to the existing version of the code (updating corners when only width/height or x/y resolution were calculated). I don't think it goes in `_update_corners_for_full_extents` since that is mostly concerned with areas that are the full extent of the projection and only deals with `x` extents at the moment.

Comment (djhoese) [reply pyresample/geometry.py:1136]: Also note the conversion in the related issue that may require other changes to this PR.

**Comment by djhoese** [comment]
@mraspaud I have some local changes that make the extents nice round numbers (usually integers, no fractional meters), but it results in a lot of the tests failing because shapes end up being off by 1 with the new logic or extents end up being slightly different. I want to run the code by you before I fix all the failing tests. What do you think:

```python
        if shape:
            height, width = shape
            x_resolution = (corners[2] - corners[0]) * 1.0 / (width - 1)
            y_resolution = (corners[3] - corners[1]) * 1.0 / (height - 1)
            area_extent = (corners[0] - x_resolution / 2,
                           corners[1] - y_resolution / 2,
                           corners[2] + x_resolution / 2,
                           corners[3] + y_resolution / 2)
        else:
            x_resolution, y_resolution = resolution
            half_x = x_resolution / 2
            half_y = y_resolution / 2
            # align extents with pixel resolution
            area_extent = (
                math.floor((corners[0] - half_x) / x_resolution) * x_resolution,
                math.floor((corners[1] - half_y) / y_resolution) * y_resolution,
                math.ceil((corners[2] + half_x) / x_resolution) * x_resolution,
                math.ceil((corners[3] + half_y) / y_resolution) * y_resolution,
            )
            width = int(np.rint((area_extent[2] - area_extent[0]) / x_resolution))
            height = int(np.rint((area_extent[3] - area_extent[1]) / y_resolution))

        return area_extent, width, height
```

The `shape` block is basically unchanged. The `else` is the important part. @yukaribbba your feedback is welcome too.

Basically I kept the width/height calculations essentially the same, but instead of adjusting the corners and then doing half-pixel math, I calculation the extents immediately and round them to the nearest pixel position. This has two benefits to the existing code:

1. I'm not adding pixels only to the right side of the area, but adding space on both sides.
2. The "round" extents mean that there is very little chance of floating point precision issues adding 0.000000005 to the resulting resolution. Especially in metered projections where we likely have no fractional parts to the extents.

**Comment by yukaribbba** [comment]
I have no problem with this. Don't forget to exclude decimals.

**Comment by djhoese** [comment]
Exclude decimals?

**Comment by yukaribbba** [comment]
Yeah I mean decimal degrees. Wait... no need for that.

**Comment by djhoese** [comment]
Yeah we should be fine since we're doing everything based on resolution.

**Comment by mraspaud** [comment]
@djhoese I'm good with that!

**Review by mraspaud — APPROVED** [review]
LGTM
